core/iwasm/common/wasm_runtime_common.h

```diff
@@ -939,6 +939,27 @@ WASM_RUNTIME_API_EXTERN bool
 wasm_runtime_unregister_natives(const char *module_name,
                                 NativeSymbol *native_symbols);
 
+/* See wasm_export.h for description */
+WASM_RUNTIME_API_EXTERN void *
+wasm_runtime_module_instance_context_key_create(
```

Comment:
The API name may be a little long and confusing? How about:
```C
wasm_runtime_create_custom_ctx_key
wasm_runtime_destroy_custom_ctx_key
wasm_runtime_set_custom_context   (or set_custom_ctx)
wasm_runtime_set_custom_context_spread
wasm_runtime_get_custom_context

wasm_native_create_custom_ctx_key
wasm_native_destroy_custom_ctx_key
wasm_native_set_custom_context   (or set_custom_ctx)
wasm_native_set_custom_context_spread
wasm_native_get_custom_context
wasm_native_call_custom_context_dtors
```

Reply:
Or wasm_runtime_create_module_isnt_ctx_key, wasm_runtime_set_module_inst_ctx?

Reply:
i agree it's long.
i'm not sure if it's confusing.

Reply:
yes, but adding prefix `wasm_runtime` really make me confusing, e.g. wasm_runtime_module_instance_set_context, how about:
```C
wasm_module_instance_create_context_key
wasm_module_instance_destroy_context_key
wasm_module_instance_set_context
wasm_module_instance_set_context_spread
wasm_module_instance_get_context
```

And using `cmake -DWAMR_BUILD_MODULE_INST_CONTEXT=1/0` and macro `WASM_ENABLE_MODULE_INST_CONTEXT` to control the feature.

The feature can be enabled no matter whether libc-wasi/libc-builtin is enabled or not.

And when libc-wasi or libc-uvwasi is enabled, the feature will be also enable automatically.

Reply:
Agree with @wenyongh

Reply:
> yes, but adding prefix `wasm_runtime` really makes me confused, e.g. wasm_runtime_module_instance_set_context, how about:

i thought we were prefixing all api with `wasm_runtime_`.

> 
> ```c
> wasm_module_instance_create_context_key
> wasm_module_instance_destroy_context_key
> wasm_module_instance_set_context
> wasm_module_instance_set_context_spread
> wasm_module_instance_get_context
> ```
> 
> And using `cmake -DWAMR_BUILD_MODULE_INST_CONTEXT=1/0` and macro `WASM_ENABLE_MODULE_INST_CONTEXT` to control the feature.

instance vs INST will make me confused.
let's be consistent.
which do you prefer?

> 
> The feature can be enabled no matter whether libc-wasi/libc-builtin is enabled or not.
> 
> And when libc-wasi or libc-uvwasi is enabled, the feature will be also enable automatically.

Reply:
> > yes, but adding prefix `wasm_runtime` really makes me confused, e.g. wasm_runtime_module_instance_set_context, how about:
> 
> i thought we were prefixing all api with `wasm_runtime_`.

Do you mean somewhat like `wasm_runtime_module_inst_create_context_key` and `wasm_runtime_module_inst_set_context`?
   
> > And using `cmake -DWAMR_BUILD_MODULE_INST_CONTEXT=1/0` and macro `WASM_ENABLE_MODULE_INST_CONTEXT` to control the feature.
> 
> instance vs INST will make me confused. let's be consistent. which do you prefer?
> 

`module_inst` in API names and `MODULE_INST` in cmake variable and compiler macro are better for me, and if you prefer using `wasm_runtime_` prefix for API:
```C
wasm_runtime_module_inst_create_context_key
wasm_runtime_module_inst_destroy_context_key
wasm_runtime_module_inst_set_context
wasm_runtime_module_inst_set_context_spread
wasm_runtime_module_inst_get_context

WAMR_BUILD_MODULE_INST_CONTEXT
WASM_ENABLE_MODULE_INST_CONTEXT
```

Not sure if it is good to others.

Reply:
> > > yes, but adding prefix `wasm_runtime` really makes me confused, e.g. wasm_runtime_module_instance_set_context, how about:
> > 
> > 
> > i thought we were prefixing all api with `wasm_runtime_`.
> 
> Do you mean somewhat like `wasm_runtime_module_inst_create_context_key` and `wasm_runtime_module_inst_set_context`?

as most of the existing apis in wasm_export.h seem to have the `wasm_runtime_` prefix,
i thought there was a project-wide policy.
it isn't my own preference.
if you prefer shorter names, it's also ok for me.

also, module_instance/module_inst part is not essential for me either.
after all, many of our api is implicitly about module instances. (eg. wasm_runtime_instantiate, wasm_runtime_set_custom_data, etc)

Reply:
> > Do you mean somewhat like `wasm_runtime_module_inst_create_context_key` and `wasm_runtime_module_inst_set_context`?
> 
> as most of the existing apis in wasm_export.h seem to have the `wasm_runtime_` prefix, i thought there was a project-wide policy. it isn't my own preference. if you prefer shorter names, it's also ok for me.
> 

Agree.

> also, module_instance/module_inst part is not essential for me either. after all, many of our api is implicitly about module instances. (eg. wasm_runtime_instantiate, wasm_runtime_set_custom_data, etc)

So you mean:
```C
wasm_runtime_create_context_key
wasm_runtime_destroy_context_key
wasm_runtime_set_context
wasm_runtime_set_context_spread
wasm_runtime_get_context
```
It is OK for me, looks concise.

@xujuntwt95329 @tonibofarull What's you opinion? Is it good to you?

Reply:
There is no problem if there is an implicit rule that `module_instance` is omitted.

Reply:
It's also OK to me, thanks